lib/SILOptimizer/Transforms/SpecializeOpaqueArchetypes.cpp

```diff
@@ -0,0 +1,348 @@
+      if (srcType.getASTType()->hasOpaqueArchetype()) {
+        src = getBuilder().createUncheckedAddrCast(
+            getOpLocation(Inst->getLoc()), src, destType);
+      } else if (destType.getASTType()->hasOpaqueArchetype()) {
+        dst = getBuilder().createUncheckedAddrCast(
+            getOpLocation(Inst->getLoc()), dst, srcType);
+      }
+    }
+    recordClonedInstruction(
+        Inst, getBuilder().createCopyAddr(getOpLocation(Inst->getLoc()), src,
+                                          dst, Inst->isTakeOfSrc(),
+                                          Inst->isInitializationOfDest()));
+  }
+
+  void visitStoreInst(StoreInst *Inst) {
+    auto src = getOpValue(Inst->getSrc());
+    auto dst = getOpValue(Inst->getDest());
+    auto srcType = src->getType();
+    auto destType = dst->getType();
+    getBuilder().setCurrentDebugScope(getOpScope(Inst->getDebugScope()));
+    // If the types mismatch cast the operands to the non opaque archetype.
+    if (destType.getASTType() != srcType.getASTType()) {
+      if (srcType.getASTType()->hasOpaqueArchetype()) {
+        assert(!srcType.isAddress());
+        src = getBuilder().createUncheckedRefCast(
+            getOpLocation(Inst->getLoc()), src, destType.getObjectType());
+      } else if (destType.getASTType()->hasOpaqueArchetype()) {
+        dst = getBuilder().createUncheckedAddrCast(
+            getOpLocation(Inst->getLoc()), dst, srcType.getAddressType());
+      }
+    }
+
+    if (!getBuilder().hasOwnership()) {
+      switch (Inst->getOwnershipQualifier()) {
+      case StoreOwnershipQualifier::Assign: {
+        auto *li = getBuilder().createLoad(getOpLocation(Inst->getLoc()), dst,
+                                           LoadOwnershipQualifier::Unqualified);
+        auto *si = getBuilder().createStore(
+            getOpLocation(Inst->getLoc()), src, getOpValue(Inst->getDest()),
+            StoreOwnershipQualifier::Unqualified);
+        getBuilder().emitDestroyValueOperation(getOpLocation(Inst->getLoc()),
+                                               li);
+        return recordClonedInstruction(Inst, si);
+      }
+      case StoreOwnershipQualifier::Init:
+      case StoreOwnershipQualifier::Trivial:
+      case StoreOwnershipQualifier::Unqualified:
+        break;
+      }
+
+      return recordClonedInstruction(
+          Inst,
+          getBuilder().createStore(getOpLocation(Inst->getLoc()), src, dst,
+                                   StoreOwnershipQualifier::Unqualified));
+    }
+
+    recordClonedInstruction(
+        Inst, getBuilder().createStore(getOpLocation(Inst->getLoc()), src, dst,
+                                       Inst->getOwnershipQualifier()));
+  }
+
+protected:
+
+  SILType remapType(SILType Ty) {
+    SILType &Sty = TypeCache[Ty];
+    if (Sty)
+      return Sty;
+
+   // Apply the opaque types substitution.
+    ReplaceOpaqueTypesWithUnderlyingTypes replacer(currentModule);
+    Sty = Ty.subst(Original.getModule(), SubsMap)
```

Comment:
since SubsMap is just the forwarding substitutions I don't think you need the first call to subst() at all, it will do nothing